### ISSUE TITLE
Poco interfaces

### ISSUE THREAD

**Issue by sjh37** [sjh37/efreversepoco#789]
Add an option to create a poco interface.

The {{moustache}} template will be something like this:

```cs
{{ClassModifier}} interface I{{NameHumanCaseWithSuffix}}{{#newline}}
{{{#newline}}

{{#each Columns}}
    {{#if OverrideModifier}}override {{/if}}{{WrapIfNullable}} {{NameHumanCase}} { get; {{PrivateSetterForComputedColumns}}set; }{{#newline}}
{{/each}}

{{#if HasReverseNavigation}}
{{#newline}}
{{#each ReverseNavigationProperty}}
    {{Definition}}{{#newline}}
{{/each}}
{{/if}}


{{#if HasForeignKey}}
{{#newline}}
{{#each ForeignKeys}}
    {{Definition}}{{#newline}}
{{/each}}
{{/if}}

{{#if EntityClassesArePartial}}
{{#newline}}
    void InitializePartial();{{#newline}}
{{/if}}
}{{#newline}}
```

**Comment by revbones-dev** [comment]
Once this is complete and the flag is set will it allow generation into separate files?

**Comment by sjh37** [comment]
Hi @revbones-duplicate Yes it will

**Comment by stevensrf11** [comment]
Does EF version 3.8.4 tt script have the ability to  have each poco entity class create be able to be derived from a custom interface?

**Comment by sjh37** [comment]
> Does EF version 3.8.4 tt script have the ability to have each poco entity class create be able to be derived from a custom interface?

@stevensrf11 Yes it does. Here is an example of adding different base classes depending on name

```cs
// Use the following function if you need to apply additional modifications to a table
// Called just before UpdateColumn
Settings.UpdateTable = delegate(Table table)
{
    // To add a base class to a table based on column names. Also see Settings.UpdateColumn below.
    var tracking  = new List<string> { "createdby", "createdon", "modifiedby", "modifiedon" };
    var replicate = new List<string> { "uniqueid", "hrid"};
    if (table.Columns.Any(x => tracking.Contains(x.NameHumanCase)))
        table.BaseClasses = " : TrackingEntity";
    if (table.Columns.Any(x => replicate.Contains(x.NameHumanCase)))
        table.BaseClasses = " : ReplicateEntity";

    // To add attributes
    table.Attributes.Add("[Serializable]");
};
```

**Comment by stevensrf11** [comment]
That example made no sense to me.

I want each poco entity class to derive IDLEntity.

 

How would I modify the tt script to do this




Sent: Friday, November 10, 2023 at 10:20 AM

**Comment by stevensrf11** [comment]
I understand this from a former post


Func<Table, string> WritePocoBaseClasses = t =>
{
    if (t.NameHumanCase== "Contact" || t.NameHumanCase== "Customer")
    {
        var r = new BaseClassMaker("BaseClass");
        return r.ToString();
    }
}

 

But I have no clue what any of that code you shown below does.

 



 

What does this do and why would I want a new list?

Does creating these new variable mess up anything or change anying that works already?

why the use of the hard coded string what do the hard code string mean?


var tracking = new List<string> { "createdby", "createdon", "modifiedby", "modifiedon" }; var replicate = new List<string> { "uniqueid", "hrid"}; var replicate = new List<string> { "uniqueid", "hrid"};

 

Is this suppose to create two base classes based for a table that  is named  by one if the string is  tracking and replicate variables?

 

if (table.Columns.Any(x => tracking.Contains(x.NameHumanCase))) table.BaseClasses = " : TrackingEntity";

if (table.Columns.Any(x => replicate.Contains(x.NameHumanCase))) table.BaseClasses = " : ReplicateEntity";


Sent: Friday, November 10, 2023 at 10:20 AM

**Comment by stevensrf11** [comment]
Ok I got it.

I see all I have to do is set table.BaseClasses equal to a class that derives from the interface.

Thanks

Steven

 
 

Sent: Friday, November 10, 2023 at 10:20 AM

**Comment by sjh37** [comment]
@stevensrf11 

```cs
Settings.UpdateTable = delegate(Table table)
{
    table.BaseClasses = " : IDLEntity";
};
```

**Comment by stevensrf11** [comment]
Thank you

**Comment by sjh37** [comment]
## The case against POCO interfaces:

EF Core entity classes are not really suited to interface abstraction. Unlike a `DbContext` (where `IDbContext` is genuinely useful for mocking/testing), entity classes are just data containers — they have no behaviour to abstract. An auto-generated `ICar` with the same properties as `Car` adds noise without value. Anyone consuming it still has to reference the concrete entity for EF queries (`DbSet<Car>`, not `DbSet<ICar>`), so the interface buys nothing in practice.

The broader problem is the maintenance burden. Every navigation property creates a circular dependency between interfaces (`ICar` references `IColour`, `IColour` references `ICar`), and nullable reference types make this messier. You end up with generated interfaces that nobody can actually use cleanly without further manual shaping.

## What people usually actually want:

When developers ask for POCO interfaces, they typically want one of two things:

1. A base class or marker interface (e.g. `IAuditable`, `ISoftDelete`, `IDLEntity`) so they can write generic repository code like `where T : IDLEntity`. This is already solved — `Settings.UpdateTable` lets you set `table.BaseClasses` to inject any base class or interface per-table or globally. That's the right hook for this.
2. Separation of concerns in Clean Architecture — entities in one project, interfaces in another. That's now solved by `UseFolderNameInNamespace` (issue #800) and will be in the next release.

## Summary:

Take a look at `Settings.UpdateTable` for the base class/marker interface use case, and note that #800 addresses the folder/namespace separation case, as that has now been completed today (15th March 2026) and will be in the next release. The full POCO interface generation as described in the template proposal is over-engineered for what it actually delivers.
